onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -281,9 +281,19 @@ class PlannerImpl {
     return elt_type->Size();
   }
 
-  static bool SameSize(const TensorShapeProto& shape1, const DataType& ptype1, const TensorShapeProto& shape2,
-                       const DataType& ptype2) {
-    return (GetElementSize(ptype1) == GetElementSize(ptype2)) && SameShape(shape1, shape2);
+  static bool SameSize(const TensorShapeProto& shape1, const onnxruntime::NodeArg& arg1, const TensorShapeProto& shape2, const onnxruntime::NodeArg& arg2) {
+    const auto& ptype1 = arg1.Type();
+    const auto& ptype2 = arg2.Type();
+    auto type1_size = GetElementSize(ptype1);
+    auto type2_size = GetElementSize(ptype2);
+    bool is_type1_string = arg1.TypeAsProto()->tensor_type().elem_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING;
+    bool is_type2_string = arg2.TypeAsProto()->tensor_type().elem_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING;
+
+    // sizeof(std::string) = sizeof(double) on gcc 4.8.x on CentOS. This causes the allocation planner to reuse
+    // a tensor of type double. This won't work for string tensors since they need to be placement new'ed.
+    // The XOR logic ensures that if one of the tensors in question is a string tensor and the other is not,
+    // we don't treat them the same when doing planning.
+    return !(is_type1_string ^ is_type2_string) && (type1_size == type2_size) && SameShape(shape1, shape2);
```

Comment:
I think it would be better to avoid reuse even if both are string, because we need to free the strings of the original tensor if we want to reuse the tensor buffer. Otherwise, we may end up with a possible memory leak of the original tensor's string contents.

Reply:
Alternatively, some logic has to be added to free the strings at the reuse-point at runtime.

Reply:
Hmm, good point. I guess it's cleaner to not allow reuse when both are strings.

Reply:
Fixed it.